### DISCUSSION TITLE
Search not working as intended

### DISCUSSION THREAD

**Discussion by xd003** [davatorium/rofi#1693] [Q&A]
Hello , this is my very first time using rofi , currently just trying out the default config in drun mode . Before i went exploring it in depth , i noticed that i have to always input minimum 3,4 characters to match/select a app which seemed quite unnecessary . For eg. With letter F , i only have Firefox app installed , still i have to always enter "fir" to match Firefox , similarly with letter S , i only have SimpleScreenRecorder , still i have to enter "sim" to match it 
I have tried all of the ```-matching methods``` but none of them seem to achieve what i need . i just need that if i enter any letter/letters , rofi should only list apps starting/matching strictly with those letter/letters & not anything else . Is there any way out to get this working the way i have mentioned ?

**Comment by DaveDavenport** [comment]
here if I type F, firefox is matched first, but your millage may vary.

Rofi matches many fields (even invisible ones) and hidden entries, you can tweak several things:
* tokenized match
* What fields to match
* what matching algorithm
* what sorting algorithm
I would try:

```
rofi -show drun -no-history -no-config -matching prefix -drun-match-fields name -no-tokenize
```

in general not disabling history will help more picked options to be placed first in the list and therefor match earlier.

**Reply by xd003** [reply]
Thank you sir , thats pretty close to what i needed , just one small change would be perfect .  i noticed while using ```rofi -show drun -no-history -no-config -matching prefix -drun-match-fields name -no-tokenize``` , rofi is prioritising apps with the search term present  anywhere in their name & not strictly sorting apps with letter in start of their name . Lemme give an example : When i search for the term S , rofi lists the apps in following order

**S**impleScreenRecorder
Rofi Theme **S**elector
nitrogen (Wallpaper **S**etter)
Avahi **S**SH Server Browser
FIle Manager **S**ettings
Qt5 **S**ettings
**S**ublime Text (Text Editor)

To put it simply , when i am searching for a letter say S , it should first match all ```^S``` apps & everything else later ,similarly searching for say ```xy``` should list all ```^xy``` apps first 

**S**impleScreenRecorder
**S**ublime Text ( Text Editor)

#Now the ones having letter "S" anywhere in their name/category could be listed
Rofi Theme **S**elector
nitrogen (Wallpaper **S**etter)
Avahi **S**SH Server Browser
FIle Manager **S**ettings
Qt5 **S**ettings

**Reply by xd003** [reply]
@DaveDavenport Although the above mentioned behavior would have ideal and most preferred . However if it is not possible ,  i can also do with just simply ignoring matches from anywhere between the names of apps so searching for S could only list the 2 apps

**Reply by DaveDavenport** [reply]
> Although the above mentioned behavior would have ideal and most preferred . 

*for you

Patches are always welcome.

**Reply by xd003** [reply]
Assuming this is the closest i can reach with current rofi to my intended functionality , i don't actually know c but i will still try it out or maybe ask someone who can help out . Can you please atleast point me to the section of code which would need changes to achieve this so i am not lost in the mix

**Reply by DaveDavenport** [reply]
Doing:
```
rofi -show drun -matching regex -no-tokenize -drun-match-fields 'name'  -drun-display-format '{name}'
```
and typing ^si is closest I think you can get.

Code for matching is here: 
https://github.com/davatorium/rofi/blob/next/source/helper.c#L222
